### ISSUE TITLE
False negatives when destructuring an array of explicit length

### ISSUE THREAD

**Issue by appsforartists** [Microsoft/TypeScript#21220]
**TypeScript Version:**  typescript@2.7.0-dev.20180116

**Code**

```ts
function takes4(a: number, b: number, c: number, d: number) {}

const has4 = [1, 2, 3, 4];
const has2 = [1, 2];

const allDestructured = takes4(...has4);
const midDestructured = takes4(1, ...has2, 3);
```

**Expected behavior:**
TypeScript knows that `has4` has 4 members and `has2` has 2 members; it can infer that `takes4`'s signature will be satisfied by either call.

**Actual behavior:**

`Expected 4 arguments, but got a minimum of { number of explicit arguments }` for each call to `takes4`.

**Related:**
I discovered this when I was trying to pass a deconstruction to [`bezier-js.Bezier`'s constructor](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/bezier-js/index.d.ts#L96).  The actual error I'm getting is that my first explicit argument (`{ x: 0, y: 0 }`) is not assignable to the type `number`.  I'm presuming that TS is choosing the wrong overload because it can't deduce that destructuring fulfills one of the overloads.

**Comment by mhegazy** [comment]
duplicate of https://github.com/Microsoft/TypeScript/issues/4130

also note that fixed length arrays (or tuples) are not inferred by default, an explicit type annotation is still required.

**Comment by appsforartists** [comment]
The code I wrote in my codebase was destructuring an argument of the type `[Point2D, Point2D]`.  I used literals in the reduction for simplicity, but I saw this bug with an explicit type too.
